README.md
# Lambda Router

A small and simple router to simplify managing routers in a NodeJS Lambda function.

**NOTE**: Prior to v2 (there was no v1) this package was owned by another maintainer ([yu0819ki](https://github.com/yu0819ki)). You can find the [old lamdba-router package here](https://github.com/yu0819ki/lambda-router). This is effectively an entirely different package.

# Installation

```
npm install lambda-router
```

# Quick Start
```javascript
const { Router } = require('lambda-router')
const router = Router({
  logger: console // uses logger-wrapper.
  includeErrorStack: process.env.stage !== 'prod'
})

router.post('/v1/endpoint', service.create)
router.get('/v1/endpoint/{id}', service.get)
router.put('/v1/endpoint/{id}', service.put)
router.delete('/v1/endpoint/{id}', service.delete)
router.unknown((event, { response }, path, method, routes) => {
  return response(404, {
    message: `You dun screwed up, now. ${path} doesn't exist!`
  })
})

async function handler (lambdaEvent, context) {
  context.callbackWaitsForEmptyEventLoop = false
  router.formatError((statusCode, error) => {
    error.errorType = getErrorType(statusCode) // custom method
    return error
  })
  let result = await router.route(lambdaEvent, context)

  return result.response
}
```

## Typescript Quickstart
```ts
// ----- app module -----
// Put these app-wide types in a module you can share with all the request handlers
// If you are using a DI kit, it should go there
// Since that's where your AppContext will be defined
import type {
  APIGatewayProxyEventV2,
  Context as LambdaContext,
} from 'aws-lambda';
import { Router } from 'lambda-router'
import type {
  RouterEvent,
  RouterContext,
  BodyResponse,
  CustomResponse,
} from 'lambda-router';


export interface Context extends LambdaContext {
  /* your custom request context data */
}
export type AppContext = RouterContext<Context>;
export type AppEvent = RouterEvent<APIGatewayProxyEventV2>;
export type RouteResponse = BodyResponse | CustomResponse;

export function initContext(event: APIGatewayProxyEventV2): Context {
  /* implement */
}
// ----- end app module -----

// ----- lambda handler module -----
const router = Router({});

async function handler(
  event: APIGatewayProxyEventV2,
  context: Context,
): Promise<APIGatewayProxyResult> {
  const appContext = initContext(event);

  const { response } = await router.route(event, appContext);

  return response;
}

// ----- end lambda handler module -----

// ----- request handler -----
type MyRouteEvent = AppEvent & {
  // You can define all requets elements here
  // including headers, querystrings, and pathParameters
  body: {
    name: string;
    age: number;
  };
};

export const routes = {
  echo: async (
    event: MyRouteEvent,
    { response }: AppContext,
  ): Promise<RouteResponse> => {
    // This is strongly typed
    const { name, age } = event.body;

    // This is also strongly typed
    return response({ statusCode: 200, body: { name, age } });
  },
};

```

# API

```javascript
function Router ({
  logger, // logger-wrapper
  assumeJson = false, // assume content-type:application/json if not content-type header is provided
  trimTrailingSlash = true, // match on paths that end with slash, e.g. "/path" matches "path/"
  extractPathParameters = true, // merge proxy path parameters into event.pathParameters
  includeTraceId = true, // include TraceId header
  inluceErrorStack = false, // include stack traces with error responses
  cors = true, // include CORS header, can be a string to set header value or true for '*'
  parseBody = true, // parse JSON or URL encoded body into event
  decodeEvent = true, // merge URI decoded parameters for path and querystring
  normalizeHeaders = false // If true all headers will be lowercased, and the original headers will be placed on event.rawHeaders
} = {}) {
  route: async (event, context) => Promise<{
    isOk,
    endpoint,
    uri,
    response
  }
  get|post|put|delete: async (pattern, handler) => {}
  unknown: (event, context, path, method, routes) => {}
  beforeRoute: async (event, context, path, method)
}
```

## Body Parsing

If `parseBody` is true and the request `Content-Type` is `application/json` or `application/x-www-form-urlencoded` the  `event.body` will be replaced with a parsed object, and `event.rawBody` will contain the original, unparsed body string.

# Routes

Routes can be registered with any of the http verb methods.

`router.[get|post|put|delete](routePattern: string, handler: (event, context) => Object|Promise)`

Route matching is done either literally or by path token. When a route is matched it's handler function is invoked, and the result is used as the response body. If a route pattern matches the path exactly it will be selected. If not, the route will try to match by token, replacing anything in curly braces with values from the path.

For example, the path `/v1/endpoint/1234` will match the route pattern `'/v1/endpoint/{id}'`, because the `{id}` section will be treated as a path token and match the value `1234`.

If the router option `extractPathParameters` is set, the `event.pathParameters` will receive the `id` value as `event.pathParameters.id: 1234`. Similarly, querystring parameters will be merged into `event.queryStringParameters`.

Route handlers that return an object will get a default status code of 200, and the object will be passed to `JSON.stringify` before being returned. Handlers that throw an error will get a default status code of 500. If you throw an error object with a `statusCode` property it's value will replace the default 500 status code. To customize status code for successful responses see the **Custom Response** section below.


## The Unknown Handler

When no route is matched the unknown handler is invoked. The default unknown handler will return a canned response containing the unmatched path, with a 404. You can replace the unknown handler by provider your own to `router.unknown`. This handler will function as a normal handler, returning a 200 unless it throws an error. Since errors default to status code 500, you should probably manually set the status code to 404.

There is a special case on the default unknown handler when the request has a match on the path but not on the method. In this case a 405 with a correct `Allow` header will be returned.

## Middleware

You can use the `beforeRoute` method to define middleware that will run before attempting to match a registered route. This is useful for mutating the incoming event or context, or for running validations at the global level:

```js
function redactAuthToken (event, context, path, method) {
  event.headers['Authorization'] = '--redacted'
}

function validateContentType (event, context, path, method) {
  if (!event.headers['Content-Type'].includes('application/json')) {
    const error = new Error('Content-Type must be JSON')
    error.statusCode = 400
    throw error
  }
}

// these will both run before any route matching occurs
router.beforeRoute(redactAuthToken)
router.beforeRoute(validateContentType)

router.post('/v1/endpoint', service.create)
router.get('/v1/endpoint/{id}', service.get)
```

# Context

The `router.route` function takes an `event` and a `context`. The `context` is useful to provide application data to the routes. It is important to know that this value must be an object, as the `route` function **mutates it** in order to provide the functionality below.

## Custom Response

The `context` object, the second parameter to route handlers, has a `response` property function. It can be used to provide custom status codes, custom headers, and control body serialization.

```javascript

router.post('/v1/endpoint', handleDelete)

async function handleCreate({ headers }, { path, response })
  let dbItem = await createNewItem()
  return response(201, result, { Location: 'https://' + headers.Host + path + '/' + dbItem.id })

async function download({ headers }, { path, response })
  let file = getFileBuffer()
  return response(201, file, { 'Content-Type': 'application/octet-stream' })
```

## Custom Headers

Fully-custom responses are sometimes undesirable, as they avoid some of the automatic behavior of normal respones (such as error formatting). If your only goal is to set a custom response header you can use `response.setHeader`

```javascript
router.post('/route', (_, { response }) => {
    response.setHeader('Location', 'something')
    return { message: 'success' }
  })
```


# Error Formatting

By default error's are serialized as responses, and either use the `error.statusCode` or a default 500 status code. If you want to control the error object, or perform custom instrumentation for errors, you can provide your own error formatter. The error formatter will be called with the status code and the error and must return the error that will be serialized for the response.

```javascript
router.formatError((statusCode, error) => {
  error.errorType = getErrorType(statusCode) // custom method
  return error
})
```

# Router response

When you call `router.route` you must provide the lambda event and context.

```javascript
let result = await router.route(event, context)
```

Optionally you can provide the the path and/or method to route on, in case the default selection method does not match your needs

```javascript
let result = await router.route(event, context, customPath, customMethod)
```

The router will match the route and return a promise that will wait on the matched route handler. The result looks like this

```javascript
{
  isOk: boolean // is the status code a 2xx
  endpoint: string // the route pattern that matched
  uri: string // the real path from the event
  response: {
    statusCode: number,
    body: string // stringified route handler response, or custom response
    headers: {} // default headers or custom response headers
  }
}

```


CHANGELOG.md
# Change Log /  Release Notes
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [3.0.0] - Unreleased
### Added
- Support for `APIGatewayProxyEventV2`
- Parse as json when `content-type: application/merge-patch+json`
### Changed
- Converted project to TypeScript
- **Breaking Change**: `trimTrailingSlash` defaults to `true`
- **Breaking Change**: `normalizeHeaders` defaults to `true`
- **Breaking Change**: `context.response` now takes an _options_ object as its only parameter
- Downgraded Jest to v26 in order to handle [this issue](https://github.com/facebook/jest/issues/12838)
### Removed
- **Breaking Change**: Support for Node < 14

## [2.12.0] - 2021-08-07
### Added
- `routes` to `unknown` as 5th parameter
- `statusCode: 405` to `defaultUnknownRoute` when other methods for the same path exist

## [2.11.0] - 2021-06-06
### Added
- `error.name` pass-through to `formatError`

## [2.10.0] - 2020-04-27
### Added
- `patch` route/verb support

## [2.9.1] - 2019-09-04
### Fixed
- Compound value `content-type` header not identifying correctly

## [2.9.0] - 2019-08-16
### Added
- `normalizeHeaders` options to lower-case all header names

## [2.8.0] - 2019-07-02
### Added
- `beforeRoute` middleware

## [2.7.0] - 2019-06-19
### Added
- `assumeJson` option

## [2.6.0] - 2019-06-19
### Added
- export `getTraceId`

## [2.5.1] - 2019-05-20
### Fixed
- Syntax error

## [2.5.0] - 2019-05-20
### Added
- Body parsing support for `application/x-www-form-urlencoded`
- `event.rawBody` as original, unparsed body when `event.body` is parsed
### Fixed
- Check `Content-Type` for `application/json` before parsing as JSON

## [2.4.1] - 2019-05-01
### Fixed
- Documentation errors

## [2.4.0] - 2018-12-01
### Fixed
- Documentation errors

## [2.3.0] - 2018-10-8
### Added
- Initial OSS release


src/router.ts
import { v4 as uuid } from 'uuid'
import { wrapper } from './logger'
import type { APIGatewayProxyEventV2, APIGatewayProxyEvent } from 'aws-lambda'

const CUSTOM_RESPONSE = Symbol('lambda-router:custom-response')
const TRACE_ID: unique symbol = Symbol('trace-id')
const METHODS = ['GET', 'POST', 'PUT', 'DELETE', 'PATCH'] as const

type Method = typeof METHODS[number]

type RequestHeaders = Record<string, string | undefined>

type ResponseHeaders = Record<string, string | boolean | number>

export type BodyResponse = object | string

export interface Response {
  statusCode: number
  headers?: ResponseHeaders
  body: string
  isBase64Encoded?: boolean
}

export interface CustomResponse extends Omit<Response, 'body'> {
  [CUSTOM_RESPONSE]: boolean
  statusCode: number
  headers?: ResponseHeaders
  isBase64Encoded?: boolean
  body?: unknown
}

export interface RouteHandler<Event, Context> {
  (event: Event, context: Context): Promise<BodyResponse | CustomResponse>
}

interface Route<Event, Context> {
  method: Method
  path: string
  handler: RouteHandler<Event, Context>
}

export type Middleware<Event, Context> = (
  event: Event,
  context: Context,
  requestPath?: string,
  httpMethod?: string
) => Promise<void>

export interface CustomResponseFn {
  (options: {
    statusCode?: number
    body?: unknown
    headers?: ResponseHeaders
    isBase64Encoded?: boolean
  }): CustomResponse
  setHeader: (header: string, value: string) => void
}

export interface HttpError extends Error {
  statusCode: number
  headers: ResponseHeaders
}

type ProxyEvent = APIGatewayProxyEvent | APIGatewayProxyEventV2

export type AddRoute<Event extends ProxyEvent, Context> = (
  path: string,
  handler: RouteHandler<RouterEvent<Event>, RouterContext<Context>>
) => void

export type RouterEvent<Event extends ProxyEvent> = Omit<Event, 'body'> & {
  body?: string | null | Record<string, unknown>
  rawBody?: string | null
  rawHeaders: RequestHeaders
}
export type RouterContext<Context> = Context & {
  includeTraceId?: boolean
  traceId?: string
  response: CustomResponseFn
}
export interface LambdaRouter<Event extends ProxyEvent, Context> {
  get: AddRoute<Event, Context>
  post: AddRoute<Event, Context>
  delete: AddRoute<Event, Context>
  put: AddRoute<Event, Context>
  patch: AddRoute<Event, Context>
  formatError: (handler: ErrorFormatter) => void
  unknown: (
    handler: (
      event: RouterEvent<Event>,
      context: RouterContext<Context>,
      path: string | undefined,
      httpMethod: string | undefined,
      routes: Array<{ path: string; method: string }>
    ) => Promise<BodyResponse | CustomResponse> | BodyResponse | CustomResponse
  ) => void
  route: (
    event: Event,
    context: Context,
    requestPath?: string,
    httpMethod?: string
  ) => Promise<RouterResponse>
  /** add middleware that receives the event & context before they are sent to route handler. Modifications are passed through */
  beforeRoute: (handler: Middleware<RouterEvent<Event>, RouterContext<Context>>) => void
}

export interface RouterResponse {
  endpoint?: string
  uri: string
  isOk: boolean
  response: Response
}

export interface RouterProps {
  logger?: any
  extractPathParameters?: boolean
  trimTrailingSlash?: boolean
  includeTraceId?: boolean
  includeErrorStack?: boolean
  cors?: boolean | string
  parseBody?: boolean
  assumeJson?: boolean
  decodeEvent?: boolean
  normalizeHeaders?: boolean
}

export interface ErrorFormatter {
  (statusCode: number, body: unknown): void
}

export function Router<Event extends ProxyEvent, Context>({
  logger: originalLogger,
  extractPathParameters = true,
  trimTrailingSlash = true,
  includeTraceId = true,
  includeErrorStack = false,
  cors = true,
  parseBody = true,
  assumeJson = false,
  decodeEvent = true,
  normalizeHeaders = true,
}: RouterProps = {}): LambdaRouter<Event, Context> {
  if (originalLogger && originalLogger.events && originalLogger.setKey) {
    originalLogger.events.on('beforeHandler', (event: Event, context: Context) => {
      originalLogger.setKey('traceId', getTraceId(event, context))
    })
  }

  const logger = wrapper(originalLogger)
  const routes: Route<RouterEvent<Event>, RouterContext<Context>>[] = []
  const addRoute = (
    method: Method,
    path: string,
    handler: RouteHandler<RouterEvent<Event>, RouterContext<Context>>
  ) => {
    routes.push({ method, path, handler })
  }

  const middleware: Middleware<RouterEvent<Event>, RouterContext<Context>>[] = []
  const addMiddleware = (handler: Middleware<RouterEvent<Event>, RouterContext<Context>>) => {
    if (typeof handler === 'function') middleware.push(handler)
  }

  let unknownRouteHandler = defaultUnknownRoute
  const defaultHeaders: ResponseHeaders = {
    'content-type': 'application/json',
  }
  if (cors) {
    defaultHeaders['access-control-allow-origin'] = typeof cors === 'string' ? cors : '*'
  }

  // External hooks
  let onErrorFormat: ErrorFormatter

  const route = async (
    lambdaEvent: Event,
    lambdaContext: Context,
    requestPath?: string,
    httpMethod?: string
  ): Promise<RouterResponse> => {
    const event: RouterEvent<Event> = {
      ...lambdaEvent,
      rawBody: lambdaEvent.body,
      rawHeaders: lambdaEvent.headers,
    }
    let statusCode, body
    const requestHeaders = normalizeRequestHeaders(event.headers)
    let headers = { ...defaultHeaders }
    // Safety Checks
    if ((lambdaContext as any).response) {
      const message =
        'context.response has already been assigned. Lambda-router reserves this property for custom responses.'
      logger.error(message)
      return Promise.reject(new Error(message))
    }
    const context = lambdaContext as RouterContext<Context>
    context.includeTraceId = includeErrorStack
    // Custom Response
    const response: any = customResponse.bind(null, context)
    response.setHeader = (header: string, value: string) => {
      headers[header] = value
    }
    context.response = response as CustomResponseFn

    // Allow method and path overrides
    httpMethod =
      httpMethod ||
      (event as unknown as APIGatewayProxyEventV2)?.requestContext?.http?.method ||
      (event as unknown as APIGatewayProxyEvent).httpMethod ||
      (event as any)?.method
    requestPath =
      requestPath ||
      (event as unknown as APIGatewayProxyEventV2)?.rawPath ||
      (event as unknown as APIGatewayProxyEvent)?.path ||
      (event as unknown as APIGatewayProxyEvent)?.requestContext?.path ||
      (event as unknown as any)?.resourcePath

    if (!httpMethod) {
      throw new Error('Unable to determine httpMethod')
    }
    if (!requestPath) {
      throw new Error('Unable to requestPath')
    }

    if (trimTrailingSlash && requestPath) {
      requestPath = requestPath.replace(/(?<=.)\/$/, '')
    }

    if (normalizeHeaders) {
      event.headers = requestHeaders
    }

    const route = getRoute(routes, event, requestPath, httpMethod, extractPathParameters)
    const hasBody = event.body && typeof event.body === 'string'
    const contentType = requestHeaders && requestHeaders['content-type']
    const jsonBody =
      hasBody &&
      (hasHeaderValue(contentType, 'application/json') ||
        hasHeaderValue(contentType, 'application/merge-patch+json') ||
        (!contentType && assumeJson))
    const urlEncodedBody =
      hasBody && hasHeaderValue(contentType, 'application/x-www-form-urlencoded')

    // Parse and decode
    try {
      if (parseBody && event.body) {
        if (jsonBody) event.body = JSON.parse(event.body as string)
        else if (urlEncodedBody)
          event.body = Object.fromEntries(new URLSearchParams(event.body as string).entries())
      }
      if (decodeEvent) {
        logger.debug('decoding parameters')
        event.pathParameters = decodeProperties(event.pathParameters || {})
        event.queryStringParameters = decodeProperties(event.queryStringParameters || {})
      }
    } catch (error: any) {
      logger.error('route error', error?.toString(), error?.stack)

      return createResponse(
        400,
        { message: 'Malformed request' },
        defaultHeaders,
        requestPath,
        requestPath
      )
    }

    // Route
    if (includeTraceId) context.traceId = headers['X-Correlation-Id'] = getTraceId(event, context)
    try {
      for (const fn of middleware) {
        await fn(event, context, requestPath, httpMethod)
      }

      const result = await (route
        ? route.handler(event, context)
        : unknownRouteHandler(event, context, requestPath, httpMethod, routes))
      if (result && (result as CustomResponse)[CUSTOM_RESPONSE] === true) {
        statusCode = (result as CustomResponse).statusCode
        body = (result as CustomResponse).body
        headers = { ...defaultHeaders, ...(result as CustomResponse).headers }
      } else {
        statusCode = 200
        body = result
      }
    } catch (error: any) {
      statusCode = error.statusCode || 500
      body = {
        ...error,
        // The spread doesn't get the non-enumerable message
        message: error.message,
        name: error.name,
        stack: includeErrorStack && error.stack,
      }
      if (error.headers) {
        headers = { ...headers, ...error.headers }
      }
      if (onErrorFormat && typeof onErrorFormat === 'function') {
        body = onErrorFormat(statusCode, body)
      }
    }

    return createResponse(statusCode, body, headers, route && route.path, requestPath)
  }

  // Bound router functions
  return {
    beforeRoute: addMiddleware,
    get: addRoute.bind(null, 'GET'),
    post: addRoute.bind(null, 'POST'),
    put: addRoute.bind(null, 'PUT'),
    delete: addRoute.bind(null, 'DELETE'),
    patch: addRoute.bind(null, 'PATCH'),
    unknown: (
      handler: (
        event: RouterEvent<Event>,
        context: RouterContext<Context>,
        path: string | undefined,
        httpMethod: string | undefined,
        routes: Array<{ path: string; method: string }>
      ) => Promise<BodyResponse | CustomResponse> | BodyResponse | CustomResponse
    ) => {
      unknownRouteHandler = handler
    },
    formatError: (handler: ErrorFormatter) => {
      onErrorFormat = handler
    },
    route,
  }
}

function customResponse<Context>(
  context: RouterContext<Context>,
  {
    statusCode = 200,
    body,
    headers = {},
    isBase64Encoded = false,
  }: {
    statusCode?: number
    body?: BodyResponse | string
    headers?: ResponseHeaders
    isBase64Encoded?: boolean
  }
): CustomResponse {
  const response = {
    [CUSTOM_RESPONSE]: true,
    statusCode,
    body,
    headers,
    isBase64Encoded,
  }
  if (context.includeTraceId && context.traceId) {
    response.headers['x-correlation-id'] = context.traceId
  }
  return response
}

function getRoute<Event extends ProxyEvent, Context>(
  routes: Route<RouterEvent<Event>, RouterContext<Context>>[] = [],
  event: RouterEvent<Event>,
  eventPath: string | undefined,
  method: string | undefined,
  tokenizePathParts: boolean
): Route<RouterEvent<Event>, RouterContext<Context>> | undefined {
  if (!eventPath || !method) return
  let route = routes.find((r) => {
    return eventPath === r.path && method === r.method
  })

  if (!route) {
    let tokens
    route = routes.find((r) => {
      if (method !== r.method) return false
      tokens = matchPathParts(eventPath, r.path)
      return !!tokens
    })
    if (tokenizePathParts && tokens) {
      if (!event.pathParameters) event.pathParameters = {}
      Object.assign(event.pathParameters, tokens)
    }
  }

  return route
}

function matchPathParts(eventPath: string, path: string): Record<string, string> | false {
  const eventPathParts = eventPath.split('/')
  const routePathParts = path.split('/')

  // Fail fast if they're not the same length
  if (eventPathParts.length !== routePathParts.length) return false
  const tokens: Record<string, string> = {}

  // Start with 1 because the url should always start with the first back slash
  for (let i = 1; i < eventPathParts.length; ++i) {
    const pathPart = eventPathParts[i]
    const routePart = routePathParts[i]

    // If the part is a curly braces value
    const pathPartMatch = /\{(\w+)}/g.exec(routePart)
    if (pathPartMatch) {
      tokens[pathPartMatch[1]] = pathPart
      continue
    }

    // Fail fast if a part doesn't match
    if (routePart !== pathPart) {
      return false
    }
  }

  return tokens
}

function defaultUnknownRoute(
  _event: any,
  _context: any,
  path: string | undefined,
  httpMethod: string | undefined,
  routes: Array<{ path: string; method: string }>
): Promise<BodyResponse | CustomResponse> | BodyResponse | CustomResponse {
  const methodMatches = routes
    .filter((r) => {
      return path === r.path
    })
    .map((r) => r.method)

  const error = new Error(`Endpoint not supported: ${httpMethod}${path}`) as HttpError
  error.statusCode = 404

  if (methodMatches.length) {
    error.statusCode = 405
    error.headers = {
      Allow: methodMatches.join(','),
    }
  }

  throw error
}

function createResponse(
  statusCode: number,
  body: BodyResponse | string,
  headers: ResponseHeaders,
  endpoint: string | undefined,
  uri: string
): RouterResponse {
  return {
    endpoint,
    uri,
    isOk: statusCode.toString()[0] === '2',
    response: createProxyResponse(statusCode, body, headers),
  }
}

export function createProxyResponse(
  statusCode: number,
  body: BodyResponse | string,
  headers: ResponseHeaders = {}
): Response {
  if (headers['content-type'] === undefined) headers['content-type'] = 'application/json'
  // output follows the format described here
  // http://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-set-up-simple-proxy.html?shortFooter=true#api-gateway-simple-proxy-for-lambda-output-format
  return {
    statusCode,
    body: typeof body === 'object' ? JSON.stringify(body) : body,
    headers: { ...headers },
  }
}

export function getTraceId(event: ProxyEvent | RouterEvent<ProxyEvent>, context: any) {
  const traceId =
    (event.headers &&
      (event.headers['X-Trace-Id'] ||
        event.headers['X-TRACE-ID'] ||
        event.headers['x-trace-id'] ||
        event.headers['X-Correlation-Id'] ||
        event.headers['X-CORRELATION-ID'] ||
        event.headers['x-correlation-id'])) ||
    context.awsRequestId ||
    uuid()
  context[TRACE_ID] = traceId
  return traceId
}

function decodeProperties(obj: Record<string, string | undefined>): Record<string, string> {
  return (
    obj &&
    Object.keys(obj).reduce((r, key) => {
      r[key] = obj[key] ? decodeURIComponent(obj[key] as string) : ''
      return r
    }, {} as Record<string, string>)
  )
}

function hasHeaderValue(header: string | undefined, value?: string): boolean {
  if (!header || !value) return false
  header = header.toLowerCase()
  value = value.toLowerCase()
  if (header === value) return true
  const headerParts = header.split(';')
  return headerParts.includes(value)
}

function normalizeRequestHeaders(reqHeaders: RequestHeaders = {}): RequestHeaders {
  return Object.keys(reqHeaders).reduce((headers, name) => {
    const reqHeader = reqHeaders[name]
    if (reqHeader) {
      headers[name.toLowerCase()] = reqHeader
    }
    return headers
  }, {} as RequestHeaders)
}
